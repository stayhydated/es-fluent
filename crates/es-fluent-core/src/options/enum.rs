use std::str::FromStr as _;

use crate::strategy::DisplayStrategy;
use bon::Builder;
use darling::{FromDeriveInput, FromField, FromMeta, FromVariant};
use getset::Getters;
use strum::IntoEnumIterator as _;

/// Options for an enum field.
#[derive(Clone, Debug, FromField, Getters)]
#[darling(attributes(fluent))]
pub struct EnumFieldOpts {
    /// The identifier of the field.
    #[getset(get = "pub")]
    ident: Option<syn::Ident>,
    /// The type of the field.
    #[getset(get = "pub")]
    ty: syn::Type,
    /// Whether to skip this field.
    #[darling(default)]
    skip: Option<bool>,
    /// Whether this field is a choice.
    #[darling(default)]
    choice: Option<bool>,
}

impl EnumFieldOpts {
    /// Returns `true` if the field should be skipped.
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }
    /// Returns `true` if the field is a choice.
    pub fn is_choice(&self) -> bool {
        self.choice.unwrap_or(false)
    }
}

/// Options for an enum variant.
#[derive(Clone, Debug, FromVariant, Getters)]
#[darling(attributes(fluent))]
pub struct VariantOpts {
    /// The identifier of the variant.
    #[getset(get = "pub")]
    ident: syn::Ident,
    fields: darling::ast::Fields<EnumFieldOpts>,
    /// Whether to skip this variant.
    #[darling(default)]
    skip: Option<bool>,
}

impl VariantOpts {
    /// Returns `true` if the variant should be skipped.
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }
    /// Returns the style of the variant's fields.
    pub fn style(&self) -> darling::ast::Style {
        self.fields.style
    }
    /// Returns the fields of the variant that are not skipped.
    pub fn fields(&self) -> Vec<&EnumFieldOpts> {
        self.fields
            .iter()
            .filter(|field| !field.is_skipped())
            .collect()
    }
    /// Returns all fields of the variant.
    pub fn all_fields(&self) -> Vec<&EnumFieldOpts> {
        self.fields.iter().collect()
    }
}

/// Options for an enum.
#[derive(Clone, FromDeriveInput, Getters, Debug)]
#[darling(supports(enum_unit, enum_named, enum_tuple), attributes(fluent))]
#[getset(get = "pub")]
pub struct EnumOpts {
    /// The identifier of the enum.
    ident: syn::Ident,
    /// The generics of the enum.
    generics: syn::Generics,
    data: darling::ast::Data<VariantOpts, darling::util::Ignored>,
    #[darling(flatten)]
    attr_args: EnumFluentAttributeArgs,
}

impl EnumOpts {
    /// Returns the variants of the enum.
    pub fn variants(&self) -> Vec<&VariantOpts> {
        match &self.data {
            darling::ast::Data::Enum(variants) => variants.iter().collect(),
            _ => unreachable!("Unexpected data type for enum"),
        }
    }
}

/// Attribute arguments for an enum.
#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
pub struct EnumFluentAttributeArgs {
    display: Option<String>,
    #[darling(default)]
    choice: Option<bool>,
    #[darling(default)]
    this: Option<bool>,
}

impl EnumFluentAttributeArgs {
    /// Returns the display strategy for the enum.
    pub fn display(&self) -> DisplayStrategy {
        if let Some(mode_str) = self.display.as_deref() {
            DisplayStrategy::from_str(mode_str).unwrap_or_else(|_| {
                let possible_values: Vec<&'static str> = DisplayStrategy::iter()
                    .map(|variant| variant.into())
                    .collect();

                panic!(
                    "Unexpected mode: '{}', expected one of {:?}",
                    mode_str, possible_values
                );
            })
        } else {
            DisplayStrategy::FluentDisplay
        }
    }
    /// Returns `true` if the enum is a choice.
    pub fn is_choice(&self) -> bool {
        self.choice.unwrap_or(false)
    }
    /// Returns `true` if the enum should be passed as `this`.
    pub fn is_this(&self) -> bool {
        self.this.unwrap_or(false)
    }
}

/// Options for an enum that can be used as a choice.
#[derive(Clone, Debug, FromDeriveInput, Getters)]
#[darling(supports(enum_unit), attributes(fluent_choice))]
#[getset(get = "pub")]
pub struct EnumChoiceOpts {
    /// The identifier of the enum.
    ident: syn::Ident,
    /// The generics of the enum.
    generics: syn::Generics,
    data: darling::ast::Data<darling::util::Ignored, darling::util::Ignored>,
    #[darling(flatten)]
    attr_args: EnumChoiceAttributeArgs,
}

/// Attribute arguments for an enum that can be used as a choice.
#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
#[getset(get = "pub")]
pub struct EnumChoiceAttributeArgs {
    #[darling(default)]
    serialize_all: Option<String>,
}
