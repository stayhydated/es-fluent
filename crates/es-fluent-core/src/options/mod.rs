pub mod r#enum;
pub mod r#struct;

const DEFAULT_FL_PATH: &str = "crate::fl";
