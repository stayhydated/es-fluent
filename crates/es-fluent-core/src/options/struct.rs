use std::str::FromStr as _;

use bon::Builder;
use darling::{FromDeriveInput, FromField, FromMeta};
use getset::Getters;
use quote::format_ident;
use strum::IntoEnumIterator as _;

use crate::strategy::DisplayStrategy;

#[derive(Clone, Debug, FromField, Getters)]
#[darling(attributes(fluent))]
pub struct StructFieldOpts {
    #[getset(get = "pub")]
    ident: Option<syn::Ident>,
    #[getset(get = "pub")]
    ty: syn::Type,
    #[darling(default)]
    skip: Option<bool>,
    #[darling(default)]
    default: Option<bool>,
}

impl StructFieldOpts {
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }

    pub fn is_default(&self) -> bool {
        self.default.unwrap_or(false)
    }
}

#[derive(Clone, FromDeriveInput, Getters)]
#[darling(supports(struct_named), attributes(fluent))]
#[getset(get = "pub")]
pub struct StructOpts {
    ident: syn::Ident,
    #[darling(default)]
    derive: darling::util::PathList,
    generics: syn::Generics,
    data: darling::ast::Data<darling::util::Ignored, StructFieldOpts>,
    #[darling(flatten)]
    attr_args: StructFluentAttributeArgs,
}

impl StructOpts {
    const FTL_ENUM_IDENT: &str = "Ftl";

    pub fn ftl_enum_ident(&self) -> syn::Ident {
        format_ident!("{}{}", &self.ident, Self::FTL_ENUM_IDENT)
    }

    pub fn keyyed_idents(&self) -> Vec<syn::Ident> {
        self.attr_args
            .clone()
            .keys
            .map_or_else(Vec::new, |keys| {
                keys.into_iter()
                    .map(|key| format_ident!("{}", key.value()))
                    .collect()
            })
            .into_iter()
            .map(|key| format_ident!("{}{}{}", &self.ident, key, Self::FTL_ENUM_IDENT))
            .collect()
    }

    pub fn fields(&self) -> Vec<&StructFieldOpts> {
        match &self.data {
            darling::ast::Data::Struct(fields) => fields
                .fields
                .iter()
                .filter(|field| !field.is_skipped())
                .collect(),
            _ => vec![],
        }
    }
}

#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
pub struct StructFluentAttributeArgs {
    #[darling(default)]
    fl: Option<syn::Path>, // this is probably useless, force user to provide a fl! macro?
    display: Option<String>,
    #[darling(default)]
    keys: Option<Vec<syn::LitStr>>,
}
impl StructFluentAttributeArgs {
    pub fn fl(&self) -> syn::Path {
        let default_fl_path: syn::Path = syn::parse_str(super::DEFAULT_FL_PATH).unwrap();
        self.fl.clone().unwrap_or(default_fl_path)
    }
    pub fn display(&self) -> DisplayStrategy {
        if let Some(mode_str) = self.display.as_deref() {
            DisplayStrategy::from_str(mode_str).unwrap_or_else(|_| {
                let possible_values: Vec<&'static str> = DisplayStrategy::iter()
                    .map(|variant| variant.into())
                    .collect();

                panic!(
                    "Unexpected mode: '{}', expected one of {:?}",
                    mode_str, possible_values
                );
            })
        } else {
            DisplayStrategy::StdDisplay
        }
    }
}
