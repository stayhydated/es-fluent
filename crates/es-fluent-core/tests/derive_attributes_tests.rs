use darling::FromDeriveInput;
use es_fluent_core::options::{
    r#enum::{EnumChoiceOpts, EnumOpts},
    r#struct::{StructKvOpts, StructOpts},
};
use syn::{DeriveInput, parse_quote};

/// EsFluent on enums: default display (fluent), no `this`, no enum-level `choice`
#[test]
fn es_fluent_enum_attributes_default_snapshot() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent] // default: display = "fluent", no this, no choice
        enum ApiError {
            NotFound,
            PermissionDenied,
            Data {
                id: u64,
                #[fluent(skip)]
                internal: bool,
            },
            TupleVariant(String, #[fluent(skip)] i32),
        }
    };

    let opts = EnumOpts::from_derive_input(&input).expect("EnumOpts should parse");
    insta::assert_debug_snapshot!(
        "es_fluent_enum_attributes_default_snapshot__analysis",
        &opts
    );
}

/// EsFluent on enums: override display to std, enable `this`, and enum-level `choice` flag
#[test]
fn es_fluent_enum_attributes_std_this_choice_snapshot() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent(display = "std", this, choice)]
        enum Status {
            // unit
            Ok,
            // tuple with a choice field and a skipped field
            Mixed(#[fluent(choice)] Severity, #[fluent(skip)] i32),
            // struct with one choice field
            Info {
                #[fluent(choice)]
                level: Severity,
                message: String,
            }
        }
    };

    let opts = EnumOpts::from_derive_input(&input).expect("EnumOpts should parse");
    insta::assert_debug_snapshot!(
        "es_fluent_enum_attributes_std_this_choice_snapshot__analysis",
        &opts
    );
}

/// EsFluent on structs: default display (fluent), with `this`, and derive list present
#[test]
fn es_fluent_struct_attributes_this_with_derive_snapshot() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent(this, derive(Debug, Clone))]
        struct Person {
            name: String,
            #[fluent(skip)]
            password_hash: String,
            #[fluent(choice)]
            gender: Gender,
        }
    };

    let opts = StructOpts::from_derive_input(&input).expect("StructOpts should parse");
    insta::assert_debug_snapshot!(
        "es_fluent_struct_attributes_this_with_derive_snapshot__opts",
        &opts
    );
}

/// EsFluent on structs: override display to std, no `this`, exercise default and choice fields
#[test]
fn es_fluent_struct_attributes_std_with_default_and_choice_snapshot() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent(display = "std")]
        struct Label {
            #[fluent(default)]
            text: String,
            #[fluent(choice)]
            style: Emphasis,
            #[fluent(skip)]
            developer_only_flag: bool,
        }
    };

    let opts = StructOpts::from_derive_input(&input).expect("StructOpts should parse");
    insta::assert_debug_snapshot!(
        "es_fluent_struct_attributes_std_with_default_and_choice_snapshot__opts",
        &opts
    );
}

/// EsFluentKv on structs: no keys provided (single FTL enum), default display is std
#[test]
fn es_fluent_kv_attributes_no_keys_snapshot() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluentKv)]
        #[fluent_kv] // default: display = "std", no keys, no this
        struct Config {
            host: String,
            port: u16,
            #[fluent_kv(skip)]
            deprecated: bool,
        }
    };

    let opts = StructKvOpts::from_derive_input(&input).expect("StructKvOpts should parse");
    insta::assert_debug_snapshot!("es_fluent_kv_attributes_no_keys_snapshot__analysis", &opts);
}

/// EsFluentKv on structs: with keys, `this`, `derive` list, and a default field
#[test]
fn es_fluent_kv_attributes_keys_this_derive_default_snapshot() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluentKv)]
        #[fluent_kv(this, keys = ["Primary", "Secondary"], derive(Debug, PartialEq))]
        struct Profile {
            #[fluent_kv(default)]
            id: u64,
            username: String,
            #[fluent_kv(skip)]
            secret_token: String,
            active: bool,
        }
    };

    let opts = StructKvOpts::from_derive_input(&input).expect("StructKvOpts should parse");
    insta::assert_debug_snapshot!(
        "es_fluent_kv_attributes_keys_this_derive_default_snapshot__analysis",
        &opts
    );
}

/// EsFluentChoice on enums: no serialize_all provided (identity)
#[test]
fn es_fluent_choice_attributes_none_snapshot() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluentChoice)]
        #[fluent_choice] // serialize_all not set
        enum Gender {
            Male,
            Female,
            Other,
        }
    };

    let opts = EnumChoiceOpts::from_derive_input(&input).expect("EnumChoiceOpts should parse");
    insta::assert_debug_snapshot!("es_fluent_choice_attributes_none_snapshot__opts", &opts);
}

/// EsFluentChoice on enums: serialize_all = "snake_case"
#[test]
fn es_fluent_choice_attributes_snake_case_snapshot() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluentChoice)]
        #[fluent_choice(serialize_all = "snake_case")]
        enum Severity {
            VeryLow,
            Low,
            Medium,
            High,
            VeryHigh,
        }
    };

    let opts = EnumChoiceOpts::from_derive_input(&input).expect("EnumChoiceOpts should parse");
    insta::assert_debug_snapshot!(
        "es_fluent_choice_attributes_snake_case_snapshot__opts",
        &opts
    );
}

/// EsFluentChoice on enums: serialize_all = "SCREAMING_SNAKE_CASE"
#[test]
fn es_fluent_choice_attributes_screaming_snake_snapshot() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluentChoice)]
        #[fluent_choice(serialize_all = "SCREAMING_SNAKE_CASE")]
        enum Emphasis {
            None,
            Light,
            Medium,
            Strong,
        }
    };

    let opts = EnumChoiceOpts::from_derive_input(&input).expect("EnumChoiceOpts should parse");
    insta::assert_debug_snapshot!(
        "es_fluent_choice_attributes_screaming_snake_snapshot__opts",
        &opts
    );
}
