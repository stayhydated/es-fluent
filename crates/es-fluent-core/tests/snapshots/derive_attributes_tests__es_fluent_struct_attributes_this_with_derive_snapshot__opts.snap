---
source: crates/es-fluent-core/tests/derive_attributes_tests.rs
expression: "&opts"
---
StructOpts {
    ident: Ident {
        sym: Person,
    },
    generics: Generics {
        lt_token: None,
        params: [],
        gt_token: None,
        where_clause: None,
    },
    data: Struct(
        Fields {
            style: Struct,
            fields: [
                StructFieldOpts {
                    ident: Some(
                        Ident {
                            sym: name,
                        },
                    ),
                    ty: Type::Path {
                        qself: None,
                        path: Path {
                            leading_colon: None,
                            segments: [
                                PathSegment {
                                    ident: Ident {
                                        sym: String,
                                    },
                                    arguments: PathArguments::None,
                                },
                            ],
                        },
                    },
                    skip: None,
                    default: None,
                    choice: None,
                },
                StructFieldOpts {
                    ident: Some(
                        Ident {
                            sym: password_hash,
                        },
                    ),
                    ty: Type::Path {
                        qself: None,
                        path: Path {
                            leading_colon: None,
                            segments: [
                                PathSegment {
                                    ident: Ident {
                                        sym: String,
                                    },
                                    arguments: PathArguments::None,
                                },
                            ],
                        },
                    },
                    skip: Some(
                        true,
                    ),
                    default: None,
                    choice: None,
                },
                StructFieldOpts {
                    ident: Some(
                        Ident {
                            sym: gender,
                        },
                    ),
                    ty: Type::Path {
                        qself: None,
                        path: Path {
                            leading_colon: None,
                            segments: [
                                PathSegment {
                                    ident: Ident {
                                        sym: Gender,
                                    },
                                    arguments: PathArguments::None,
                                },
                            ],
                        },
                    },
                    skip: None,
                    default: None,
                    choice: Some(
                        true,
                    ),
                },
            ],
            span: Some(
                bytes(0..0),
            ),
            __nonexhaustive: (),
        },
    ),
    attr_args: StructFluentAttributeArgs {
        display: None,
        this: Some(
            true,
        ),
        derive: PathList(
            [
                Path {
                    leading_colon: None,
                    segments: [
                        PathSegment {
                            ident: Ident {
                                sym: Debug,
                            },
                            arguments: PathArguments::None,
                        },
                    ],
                },
                Path {
                    leading_colon: None,
                    segments: [
                        PathSegment {
                            ident: Ident {
                                sym: Clone,
                            },
                            arguments: PathArguments::None,
                        },
                    ],
                },
            ],
        ),
    },
}
